src/embedding/kd.ts

```diff
@@ -0,0 +1,752 @@
+         * Type represent that the platform is down.
+         */
+        Platform = 'platform'
+    }
+
+    /**
+     * List of available commands. This is included in each event sent by KD.
+     */
+    export interface IKdAvailableCommands {
+        /**
+         * Array of avaiable commands types.
+         */
+        availableCommands: GdcKdCommandType[];
+    }
+
+    /**
+     * This event will be emitted if KD runs into errors while processing the posted command.
+     *
+     * @remarks see {@link GdcErrorType} for types of errors that may fly
+     */
+    export type KdCommandFailed = CommandFailed<GdcErrorType>;
+
+    /**
+     * Base type for the data of error events sent by AD
+     * in case command processing comes to an expected or unexpected halt.
+     */
+    export type KdCommandFailedData = CommandFailedData<GdcProductName.KPI_DASHBOARD>;
+
+    /**
+     * Type-guard checking whether an object is an instance of {@link KdCommandFailedData}
+     *
+     * @param obj - object to test
+     */
+    export function isAdCommandFailedData(obj: any): obj is KdCommandFailedData {
+        return isCommandFailedData<GdcProductName.KPI_DASHBOARD>(obj);
+    }
+
+    /**
+     * Save command body sent by outer application
+     */
+    export interface IKdSaveCommandBody {
+        /**
+         * Dashboard title - use as title of new dashboard or rename of saved dashboard
+         */
+        title: string;
+    }
+
+    /**
+     * Saves current dashboard.
+     *
+     * Contract:
+     *
+     * -  if currently edited dashboard IS NOT eligible for save (empty, in-error), then KdCommandFailed event
+     *    will be posted
+     * -  if the specified title is invalid / does not match title validation rules, then KdCommandFailed event
+     *    will be posted
+     * -  otherwise dashboard WILL be saved with the title as specified in the body and the DashboardSaved event
+     *    will be posted
+     * -  the DashboardSaved event will be posted even when saving dashboard that has not changed but would
+     *    otherwise be eligible for saving (not empty, not in-error)
+     *
+     * Note: sending Save command with different title means dashboard will be saved with that new title.
+     *
+     * @remarks use {@link SaveDashboardCommand} factory function to instantiate
+     */
+    export type SaveDashboardCommand = IKdMessageEvent<GdcKdCommandType.Save, IKdSaveCommandBody>;
+
+    export type SaveDashboardCommandData = IKdGdcMessageEnvelope<GdcKdCommandType.Save, IKdSaveCommandBody>;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link SaveDashboardCommandData}.
+     *
+     * @param obj - object to test
+     */
+    export function isSaveDashboardCommandData(obj: any): obj is SaveDashboardCommandData {
+        return getEventType(obj) === GdcKdCommandType.Save;
+    }
+
+    /**
+     * It's main content is empty.
+     */
+    export type CancelEditBody = IKdAvailableCommands;
```

Comment:
Please re-scan any types here referring to `IKdAvailableCommands ` with empty body. It could not be `IKdAvailableCommands`

You can try `export type CancelEditCommand = IKdMessageEvent<GdcKdCommandType.CancelEdit, null>;`

Reply:
Cc @lanhq , could you give some advice here?

Reply:
updated